.travis.yml

```diff
@@ -1,39 +1,83 @@
 language: cpp
+dist: trusty
 
-compiler:
-  - gcc
-  - clang
+matrix:
+  include:
+    - name: "GCC 4.9 (Linux)"
+      os: linux
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - g++-4.9
+            - ninja-build
+      env:
+         - MATRIX_EVAL="export CC=gcc-4.9 && export CXX=g++-4.9"
 
-os:
-  - linux
-  - osx
+    - name: "GCC 9 (Linux)"
+      os: linux
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - g++-9
+            - ninja-build
+      env:
+        - MATRIX_EVAL="export CC=gcc-9 && export CXX=g++-9"
 
-dist: trusty
+    - name: "Clang 5.0 (Linux)"
+      os: linux
+      addons:
+        apt:
+          sources:
+            - llvm-toolchain-trusty-5.0
+          packages:
+            - clang-5.0
+            - ninja-build
+      env:
+        - MATRIX_EVAL="export CC=clang-5.0 && export CXX=clang++-5.0"
 
-addons:
-  apt:
-    packages:
-      clang
+    - name: "Clang Xcode 7.3 (Mac)"
+      os: osx
+      osx_image: xcode7.3
+      addons:
+        homebrew:
+          packages:
+            - ninja
+          update: true
+
+    - name: "Clang Xcode 11 (Mac)"
+      os: osx
+      osx_image: xcode11
+      addons:
+        homebrew:
+          packages:
+            - ninja
 
 # NOTE: Currently I commented out testing of python binding on OSX because they do not support
 # Python building and I have to provide the tools myself. I could probably install correct pip
 # (and maybe python) version myself (homebrew?), however I haven't gotten to that yet.
 
+before_install:
+  - eval "${MATRIX_EVAL}"
+
 install:
   - if [ $TRAVIS_OS_NAME == "linux" ]; then sudo apt-get install valgrind; fi
   - if [ $TRAVIS_OS_NAME == "linux" ]; then sudo -H pip install cython; fi  # Needed to build Python module.
 
 before_script:
   # Build C/C++ library and apps.
-  - mkdir -p build && cd build && CXXFLAGS="-Werror" cmake -DCMAKE_VERBOSE_MAKEFILE=ON .. && make && cd ..
+  - mkdir -p build && cd build && CXXFLAGS="-Werror" cmake -GNinja .. && ninja -v && cd ..
```

Comment:
Why ninja instead of make? I know that ninja is somewhat faster, but make isn't make more ubiquitous / standard? And I don't believe it really makes any difference regarding the speed?

Reply:
CMake's `make` is solid and production-grade, but produces a ton of crap in its logs. Using `ninja`

1. Tests that CMake is written properly (for instance, using ninja requires CMake to tell ninja what all outputs of all rules is, otherwise it might having missing edges in the DAG) and finds more cases of edgecases in your definitions
2. Produces **much** cleaner logs.

Alternatively, we could say, use `ninja` on Linux and `make` on macOS and test both code paths?

Reply:
Ok, I believe you have more knowledge regarding this topic so I will leave the decision to you. My main thought is to go for simplicity / easy maintenance when possible, and if ninja makes that slightly more complicated to bring small benefits, is it worth it? But as I said, I believe you can judge that better than me, so I will leave it to you.
Why not use ninja everywhere there, why would you go with make on OSX?

Reply:
Here's an example:
https://bugs.gentoo.org/638076

Supertux is a Mario Kart-style clone for Linux. In Gentoo, we try to build all CMake-based packages using `ninja` as a first choice, due to near-perfect scaling of its non-recursive nature (CMake Makefiles are **always** recursive, which is extremely annoying and doesn't scale. Think about people building software with `-j32` or something). The build failed with ninja. As it turns out, the CMake build system was incomplete: there is one target, and it didn't tell CMake that as a byproduct (output) of its invocation, it would produce the missing `libsqstdlib_static.a` archive. In this case, the legacy Makefile code generation backend masked a missing edge in the build DAG. Had the upstream used `ninja` in their CI, they would have instantly caught the bug.

On the other hand, I have never in my life encountered a CMake-based package where `ninja` worked, but the Makefile backend failed.

Reply:
My suggestion to use `make` on macOS was to try and cover both CMake codepaths. Personally, I would just test `ninja`, because I've never seen `ninja` work but `make` not with CMake.

Reply:
@SoapZA sounds good, let's go with ninja then!